### DISCUSSION TITLE
Consider redesign or better documentation of HDF5 library initialization for better compatibility with MPI

### DISCUSSION THREAD

**Discussion by jhendersonHDF** [HDFGroup/hdf5#3533] [Proposed Change]
In a parallel HDF5 C application, certain actions, such as assigning a variable to H5T_NATIVE_INT, can cause the library to be initialized before MPI is initialized by a call to MPI_Init within the application. This has the side effect of causing HDF5 to not properly register parallel teardown routines with the MPI implementation and can lead to issues with MPI routines being used after MPI_Finalize has been called in the application as HDF5 closes remaining IDs that are open.

To deal with this issue, either a change to, or better documentation of, HDF5 library initialization will be needed. One possible solution for the former case would be to require an explicit call to H5open to initialize HDF5.

**Comment by wrwilliams** [comment]
Are there any thoughts on the HDF5 side about migrating the MPI backend to use sessions and thus control the lifetime of HDF5's MPI usage independently from the rest of the application?
